streams/src/main/java/org/apache/kafka/streams/processor/internals/SourceNode.java

```diff
@@ -28,28 +30,32 @@
     private final List<String> topics;
 
     private ProcessorContext context;
-    private Deserializer<K> keyDeserializer;
-    private Deserializer<V> valDeserializer;
+    private ExtendedDeserializer<K> keyDeserializer;
+    private ExtendedDeserializer<V> valDeserializer;
     private final TimestampExtractor timestampExtractor;
 
     public SourceNode(String name, List<String> topics, TimestampExtractor timestampExtractor, Deserializer<K> keyDeserializer, Deserializer<V> valDeserializer) {
         super(name);
         this.topics = topics;
         this.timestampExtractor = timestampExtractor;
-        this.keyDeserializer = keyDeserializer;
-        this.valDeserializer = valDeserializer;
+        this.keyDeserializer = keyDeserializer != null ? ensureExtended(keyDeserializer) : null;
+        this.valDeserializer = valDeserializer != null ? ensureExtended(valDeserializer) : null;
     }
 
     public SourceNode(String name, List<String> topics, Deserializer<K> keyDeserializer, Deserializer<V> valDeserializer) {
         this(name, topics, null, keyDeserializer, valDeserializer);
     }
 
-    K deserializeKey(String topic, byte[] data) {
-        return keyDeserializer.deserialize(topic, data);
+    private <T> ExtendedDeserializer<T> ensureExtended(Deserializer<T> deserializer) {
```

Comment:
It looks like we have this bit of code in three different places now. Once in `Fetcher` and twice in this PR. It might make sense to have a utility function in `org.apache.kafka.common.serialization` that does this rather than repeating it. Thoughts?